File: Stage32/Python/Problem2618.py
import sys
sys.setrecursionlimit(10**6)

input = sys.stdin.readline

def main():
    N = int(input())
    W = int(input())

    event = [(0, 0)] * (W + 1)
    dp = [[-1] * (W + 1) for _ in range(W + 1)]
    path = [[0] * (W + 1) for _ in range(W + 1)]

    for i in range(1, W + 1):
        x, y = map(int, input().split())
        event[i] = (x, y)

    def dist(frm, to, type_):
        if frm == 0:
            if type_ == 1:
                x1, y1 = 1, 1
            else:
                x1, y1 = N, N
        else:
            x1, y1 = event[frm]

        x2, y2 = event[to]
        return abs(x1 - x2) + abs(y1 - y2)

    def solve(a, b):
        if a == W or b == W:
            return 0

        if dp[a][b] != -1:
            return dp[a][b]

        nxt = max(a, b) + 1

        case1 = solve(nxt, b) + dist(a, nxt, 1)
        case2 = solve(a, nxt) + dist(b, nxt, 2)

        if case1 < case2:
            path[a][b] = 1
            dp[a][b] = case1
        else:
            path[a][b] = 2
            dp[a][b] = case2

        return dp[a][b]

    def trace(a, b):
        if a == W or b == W:
            return

        nxt = max(a, b) + 1

        if path[a][b] == 1:
            print(1)
            trace(nxt, b)
        else:
            print(2)
            trace(a, nxt)

    print(solve(0, 0))
    trace(0, 0)

if __name__ == "__main__":
    main()

File: Stage32/C/Problem2618.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX 1001

int N, W;
int event[MAX][2];
int dp[MAX][MAX];
int path[MAX][MAX];

int abs_val(int x) {
    return x < 0 ? -x : x;
}

int dist(int from, int to, int type) {
    int x1, y1;

    if (from == 0) {
        if (type == 1) {
            x1 = 1; y1 = 1;
        } else {
            x1 = N; y1 = N;
        }
    } else {
        x1 = event[from][0];
        y1 = event[from][1];
    }

    int x2 = event[to][0];
    int y2 = event[to][1];

    return abs_val(x1 - x2) + abs_val(y1 - y2);
}

int solve(int a, int b) {
    if (a == W || b == W) return 0;

    if (dp[a][b] != -1) return dp[a][b];

    int next = (a > b ? a : b) + 1;

    int case1 = solve(next, b) + dist(a, next, 1);
    int case2 = solve(a, next) + dist(b, next, 2);

    if (case1 < case2) {
        path[a][b] = 1;
        return dp[a][b] = case1;
    } else {
        path[a][b] = 2;
        return dp[a][b] = case2;
    }
}

void trace(int a, int b) {
    if (a == W || b == W) return;

    int next = (a > b ? a : b) + 1;

    if (path[a][b] == 1) {
        printf("1\n");
        trace(next, b);
    } else {
        printf("2\n");
        trace(a, next);
    }
}

int main() {
    scanf("%d", &N);
    scanf("%d", &W);

    for (int i = 0; i <= W; i++) {
        for (int j = 0; j <= W; j++) {
            dp[i][j] = -1;
        }
    }

    for (int i = 1; i <= W; i++) {
        scanf("%d %d", &event[i][0], &event[i][1]);
    }

    printf("%d\n", solve(0, 0));
    trace(0, 0);

    return 0;
}